src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,153 @@
+                                      .value_or(PipelineGenerator{})} {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    // User can specify any pipeline here, but it's worth noting that some will not play nicely. For
+    // example, adding a $match will deflate the sample size.
+    PipelineGenerator pipelineSuffixGenerator;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(debug) << "Beginning to run SamplingLoader";
+            // Read the sample size documents into a vector.
+            mongocxx::pipeline samplePipeline;
+            samplePipeline.sample(config->sampleSize);
+            samplePipeline.project(make_document(kvp("_id", 0)));
+            const auto suffixStages =
+                pipeline_helpers::makePipeline(config->pipelineSuffixGenerator).view_array();
+            samplePipeline.append_stages(suffixStages);
+            auto cursor =
+                config->collection.aggregate(samplePipeline, mongocxx::options::aggregate{});
+            const auto sampleDocs =
+                std::vector<bsoncxx::document::value>(cursor.begin(), cursor.end());
+
+            if (sampleDocs.empty()) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "Sample was unable to find any documents from collection '" +
+                    to_string(config->collection.name()) +
+                    "'. Could the collection be empty or could the pipeline be filtering out "
+                    "documents? Attempting to sample " +
+                    boost::to_string(config->sampleSize) +
+                    " documents. Pipeline suffix = " + to_json(suffixStages)));
+            }
+            if (sampleDocs.size() < config->sampleSize) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "Could not get a sample of the expected size. Either the collection '" +
+                    to_string(config->collection.name()) +
+                    "' is smaller than the requested sample size of " +
+                    boost::to_string(config->sampleSize) +
+                    " documents, or the specified pipeline suffix is filtering documents. Found "
+                    "only " +
+                    boost::to_string(sampleDocs.size()) +
+                    " documents. Pipeline suffix = " + to_json(suffixStages)));
+            }
+
+            // Maintain an index into the sample that's used for the insert batches so we
+            // insert roughly the same number of copies for each sample.
+            size_t sampleIdx = 0;
+
+            // We'll re-use the same vector slots each time, just overwriting the memory on each new
+            // batch.
+            auto batchOfDocs = std::vector<bsoncxx::document::view_or_value>{
```

Comment:
I thought `std::vector<T> v{n};` would create a vector with a single element "n" and to create a vector of size n with default-initialized values (that is, T()) would need to do `std::vector<T> v(n);`, no?

Reply:
Hmm no this construction appears to be working correctly - I don't see how it could make a vector with a single element of BSON, when there is no way to construct a bson object from a number. I tried messing around with the examples here and it also seems to work for a vector of strings: https://en.cppreference.com/w/cpp/container/vector/vector

Reply:
Interesting. The compiler chooses either list initialization or direct initialization depending on whether there is an implicit conversion available for the type. Brrr. My preference would be to avoid {single_int} constructors for vectors just to stay clear of this smartness.
```
	struct Foo {
		Foo(int n) : val(n, 'a') {}
		string val;
	};
	struct Bar {
		Bar() {}
		explicit Bar(int n) : val(n, 'a') {}
		string val;
	};
	TEST_METHOD(experiment) {
		vector<Foo> v_foo{ 3 };
		Assert::AreEqual<size_t>(1, v_foo.size());

		vector<Bar> v_bar{ 3 };
		Assert::AreEqual<size_t>(3, v_bar.size());
	}
```
A mute point anyway now.